src/de/raw.rs

```diff
@@ -511,14 +511,16 @@ impl<'d, 'de> DocumentAccess<'d, 'de> {
         F: FnOnce(&mut Self) -> Result<O>,
     {
         let start_bytes = self.root_deserializer.bytes.bytes_read();
-        let out = f(self);
+        let out = f(self)?;
         let bytes_read = self.root_deserializer.bytes.bytes_read() - start_bytes;
-        *self.length_remaining -= bytes_read as i32;
-
-        if *self.length_remaining < 0 {
+        let bytes_read: i32 = bytes_read
+            .try_into()
+            .map_err(|_| Error::custom("overflow in read size"))?;
+        if bytes_read > *self.length_remaining {
```

Comment:
Testing before the subtraction means there's no chance for overflow; this change alone would also have fixed the problem, but would have masked the original error.  I included both for robustness.